compiler/src/iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensorsViaRegionOps.cpp

```diff
@@ -598,10 +598,14 @@ static FailureOr<SmallVector<Flow::DispatchWorkgroupsOp>> createFusionGroups(
   // TODO: Incrementally add ops to an empty DispatchGroupOp instead of
   // annotating fusion group IDs via attributes.
   funcOp.walk([&](Operation *op) {
-    if (hasRootOpAttribute(op)) roots[getRootNumber(op)] = op;
+    if (hasRootOpAttribute(op)) {
```

Comment:
I am still trying to weight if we want an explicit in-memory representation of `producers`. Thats is a big data structure that mirrors effectively the whole problem. A transient attribute is cheaper?

Reply:
The current implementation has 2 phases:
1. Run the heuristic
2. Rewrite the IR

The attribute approach makes is easy to build heuristics. (And easier to debug, because we can dump the IR with the attributes.) In-between step 1 and step 2, we can dump the IR to debug/experiment with the heuristic.

Once we enter phase 2, it is cheaper to have an in-memory representation. For each root op, I need to get a handle to all producers so that I can move them into the dispatch region. Without an in-memory representation, that is one walk over the entire IR per root op. With an in-memory representation, it is just one walk.

The `producers` DenseMap could easily be replaced with a `SmallVector<Operation *> getProducers(Operation *rootOp)` helper function that does the walk for each root op.

Reply:
Clarification: The commit message `Remove attributes in createFusionGroups` means "Remove the attributes at the end of createFusionGroups instead of at end of runOnOperation". It does not mean "No longer use attributes".